### ISSUE TITLE
invalid input syntax for type timestamp with time zone

### ISSUE THREAD

**Issue by koreiko** [pgpartman/pg_partman#301]
**OS Version:** "PostgreSQL 11.6 (Ubuntu 11.6-1.pgdg18.04+1) on x86_64-pc-linux-gnu, compiled by gcc (Ubuntu 7.4.0-1ubuntu1~18.04.1) 7.4.0, 64-bit"
**pg_partman version:** 4.2.2

**commands runned:**
CREATE SCHEMA partman;
CREATE EXTENSION pg_partman SCHEMA partman;

CREATE TABLE public.argon_messages_test
(
    serial_number bigint NOT NULL,
    serial_timestamp timestamp without time zone NOT NULL DEFAULT now(),
    type integer NOT NULL,
    sub_type integer NOT NULL,
    payload_version integer NOT NULL,
    payload bytea NOT NULL,
    raw_data bytea NOT NULL,
    date_created timestamp without time zone NOT NULL DEFAULT now(),
    date_modify timestamp(6) without time zone,
    CONSTRAINT argon_messages_pkey_test PRIMARY KEY (serial_number, serial_timestamp),
    CONSTRAINT positive_type_test CHECK (type > 0),
    CONSTRAINT positive_subtype_test CHECK (sub_type > 0),
    CONSTRAINT positive_payloadversion_test CHECK (payload_version > 0)
) PARTITION BY RANGE (serial_timestamp);

SELECT partman.create_parent(
    p_parent_table := 'public.argon_messages_test'::text, 
    p_control := 'serial_timestamp'::text,      			
    p_type := 'native'::text,
    p_interval := 'daily'::text,                  			
    p_constraint_cols := NULL,                      		
    p_premake := 10::int,                            		
    p_automatic_maintenance := 'on'::text,          		
    p_start_partition := ''::text,                  		
    p_inherit_fk := true,                           		
    p_epoch := 'none'::text,                        		
    p_upsert := ''::text,                           		
    p_publications := NULL::text[],                 		
    p_trigger_return_null := true,                  		
    p_template_table := NULL::text,                 		
    p_jobmon := true,                               		
    p_debug := true) ; 

**Error raised:**
ERROR:  invalid input syntax for type timestamp with time zone: ""
CONTEXT: PL/pgSQL function partman.create_parent(text,text,text,text,text[],integer,text,text,boolean,text,text,text[],boolean,text,boolean,boolean) line 367 at assignment
DETAIL: 
HINT: 
CONTEXT:  PL/pgSQL function partman.create_parent(text,text,text,text,text[],integer,text,text,boolean,text,text,text[],boolean,text,boolean,boolean) line 760 at RAISE
SQL state: P0001

**Comment by keithf4** [comment]
pg_partman currently only works with the data type `timestamp with time zone`. Unless you have a very specific need for `timestamp without time zone`, I would highly encourage you to change your datatype. Even then, you may want to rethink your need unless you can 100% guarantee that all systems that will ever talk to this database are the exact same timezone and will be so forever. If not, then you will find your time-based data giving very unexpected results since you will have no context as to what the timezone was at the time the data was inserted.

Some more concrete examples are here: https://www.depesz.com/2014/04/04/how-to-deal-with-timestamps/

**Comment by jeremybradbury** [comment]
@keithf4 I have a hard disagree with this notion:

> Even then, you may want to rethink your need unless you can 100% guarantee that all systems that will ever talk to this database are the exact same timezone and will be so forever. If not, then you will find your time-based data giving very unexpected results since you will have no context as to what the timezone was at the time the data was inserted.

1. dates should *all* be "saved as UTC+0/GMT" no matter which timezone it comes from: 
 - `new Date(any parable date or null for now).getTime()` or 
 - `new Date(any parable date or null for now).toISOString()` 
 - will get the raw data which the browser stores in GMT/UTC-0.
2. when the GMT date arrives in the browser: `new Date(some date in GMT)`.

the date is stored in TZ 0, then when the browser *displays* the date. it's magically parsed into the browsers default.

so then after any edits in the user's local time zone, you swap back step 2 and save it in the single unified timezone your company stores.

the timezone then becomes an illusion or calculation done in the client only so everyone can share a clock and all the math is transparent even to the developer (unless not using a browser, but this is still able to be cleanly done across devices).

I would honestly rethink how you choose to handle dates from an enterprise architecture pattern. 

TLDR: the servers and databases should always be set in UTC+0 served out in that same universal timezone, then the client handles adding and subtracting hours based on it's locale.

**Comment by keithf4** [comment]
I'm not sure what this response is for. The `timestamp without time zone` data type specifically does NOT store the timestamp in UTC. It stores the time value exactly as given by the client and is why I'm recommending against it and why pg_partman does not support it. Please see the blog post by Depesz that I reference. The part that you are quoting is saying that, even if they think they can provide a valid use-case for `timestamp without time zone`, they likely still aren't considering all the consequences.

While it would be nice for everyone to set their servers to UTC, and I would highly encourage it myself as well, it's not always possible and using the proper data type in PostgreSQL can help with that.
